File: Code.cpp
/*
 * Rosolino Mangano
 * CIS 3100
 * Professor Jairam
 * 05/22/2024
 */

#include <iostream> // For input and output stream operations (cin, cout)
#include <fstream>  // For file handling (ifstream)
#include <sstream>  // For string stream operations (istringstream)
#include <vector>   // For using the vector container
#include <string>   // For using the string class
#include <algorithm>// For algorithms like random_shuffle and max_element
#include <cmath>    // For mathematical functions (sqrt, exp, pow)
#include <map>      // For using the map container
#include <set>      // For using the set container
#include <utility> // For using std::pair
#include <cstdlib> // For exit() and EXIT_FAILURE functions

using namespace std;

// Helper function to split a string by a delimiter
vector<string> split(const string& s, char delimiter) {
    vector<string> tokens;
    string token;
    istringstream tokenStream(s);
    while (getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}

// DataHandler class to handle reading CSV file, splitting data into train/test sets, and separating features and labels
class DataHandler {
public:
    vector<vector<float> > features;
    vector<string> labels;

    DataHandler(const string& filename) { // Constructor takes filename
        read_csv(filename);
    }

    void read_csv(const string& filename) {
        ifstream file(filename);
        string line;
        if (!file.is_open()) {
            cerr << "Failed to open file: " << filename << endl;
            exit(EXIT_FAILURE);
        }

        // Skip the header row
        if (getline(file, line)) {
            // Verify that the header row matches expected columns
            vector<string> headers = split(line, ',');
            cout << "Headers: ";
            for (size_t i = 0; i < headers.size(); ++i) {
                cout << headers[i] << " ";
            }
            cout << endl;
        }

        while (getline(file, line)) {
            vector<string> tokens = split(line, ',');
            vector<float> row_features;
            for (size_t i = 0; i < tokens.size() - 1; ++i) {
                try {
                    row_features.push_back(stof(tokens[i]));
                } catch (const invalid_argument& e) {
                    cerr << "Invalid argument: " << e.what() << " at line: " << line << endl;
                    exit(EXIT_FAILURE);
                } catch (const out_of_range& e) {
                    cerr << "Out of range error: " << e.what() << " at line: " << line << endl;
                    exit(EXIT_FAILURE);
                }
            }
            features.push_back(row_features);
            labels.push_back(tokens.back());
        }
        file.close();
    }
};

// NaiveBayesClassifier class to implement Naive Bayes algorithm from scratch
class NaiveBayesClassifier {
public:
    map<string, vector<float> > means; // Container to hold the means of features for each class
    map<string, vector<float> > stds; // Container to hold the standard deviations of features for each class
    map<string, float> class_probabilities; // Container to hold the probabilities of each class

    // Method to train the classifier
    void fit(const vector<vector<float> >& X, const vector<string>& y) {
        calculate_class_probabilities(y); // Calculate class probabilities
        calculate_means_stds(X, y); // Calculate means and standard deviations
    }

    // Method to calculate the probability of each class based on label frequency
    void calculate_class_probabilities(const vector<string>& y) {
        map<string, int> class_counts;
        for (vector<string>::const_iterator label = y.begin(); label != y.end(); ++label) {
            class_counts[*label]++;
        }
        int total_count = y.size();
        for (map<string, int>::const_iterator pair = class_counts.begin(); pair != class_counts.end(); ++pair) {
            class_probabilities[pair->first] = static_cast<float>(pair->second) / total_count;
        }
    }

    // Method to calculate the mean and standard deviation for each class and each feature
    void calculate_means_stds(const vector<vector<float> >& X, const vector<string>& y) {
        map<string, vector<vector<float> > > label_features;
        for (size_t i = 0; i < y.size(); ++i) {
            label_features[y[i]].push_back(X[i]);
        }
        for (map<string, vector<vector<float> > >::const_iterator pair = label_features.begin(); pair != label_features.end(); ++pair) {
            string label = pair->first;
            const vector<vector<float> > &features = pair->second;
            int n_features = features[0].size();
            vector<float> mean(n_features, 0.0);
            vector<float> std(n_features, 0.0);
            for (vector<vector<float> >::const_iterator feature_set = features.begin(); feature_set != features.end(); ++feature_set) {
                for (int i = 0; i < n_features; ++i) {
                    mean[i] += (*feature_set)[i];
                }
            }
            for (int i = 0; i < n_features; ++i) {
                mean[i] /= features.size();
            }
            for (vector<vector<float> >::const_iterator feature_set = features.begin(); feature_set != features.end(); ++feature_set) {
                for (int i = 0; i < n_features; ++i) {
                    std[i] += pow((*feature_set)[i] - mean[i], 2);
                }
            }
            for (int i = 0; i < n_features; ++i) {
                std[i] = sqrt(std[i] / features.size());
            }
            means[label] = mean;
            stds[label] = std;
        }
    }

    // Method to predict the class of a single feature set
    string predict_single(const vector<float>& input_features) {
        map<string, float> probabilities;
        for (map<string, vector<float> >::const_iterator pair = means.begin(); pair != means.end(); ++pair) {
            string label = pair->first;
            probabilities[label] = class_probabilities[label];
            for (size_t i = 0; i < input_features.size(); ++i) {
                probabilities[label] *= calculate_probability(input_features[i], means[label][i], stds[label][i]);
            }
        }

        // Find the label with the maximum probability
        string max_label;
        float max_prob = -1.0;
        for (map<string, float>::const_iterator prob = probabilities.begin(); prob != probabilities.end(); ++prob) {
            if (prob->second > max_prob) {
                max_prob = prob->second;
                max_label = prob->first;
            }
        }

        return max_label;
    }

    // Method to calculate the probability of a feature value with a Gaussian distribution
    float calculate_probability(float x, float mean, float std) {
        float exponent = exp(-pow(x - mean, 2) / (2 * pow(std, 2)));
        return (1 / (sqrt(2 * M_PI) * std)) * exponent;
    }

    // Method to predict a list of feature sets
    vector<string> predict(const vector<vector<float> >& X) {
        vector<string> predictions;
        for (vector<vector<float> >::const_iterator features = X.begin(); features != X.end(); ++features) {
            predictions.push_back(predict_single(*features));
        }
        return predictions;
    }

    // Method to calculate the accuracy of the predictions
    float accuracy(const vector<string>& y_test, const vector<string>& y_pred) {
        int correct = 0;
        for (size_t i = 0; i < y_test.size(); ++i) {
            if (y_test[i] == y_pred[i]) {
                correct++;
            }
        }
        return static_cast<float>(correct) / y_test.size();
    }

    // Method to generate a classification report for the predictions
    map<string, map<string, float> > classification_report(const vector<string>& y_true, const vector<string>& y_pred) {
        set<string> unique_labels(y_true.begin(), y_true.end());
        map<string, map<string, float> > report;
        for (set<string>::const_iterator label = unique_labels.begin(); label != unique_labels.end(); ++label) {
            int tp = 0, fp = 0, fn = 0, tn = 0;
            for (size_t i = 0; i < y_true.size(); ++i) {
                if (y_true[i] == *label && y_pred[i] == *label) {
                    tp++;
                } else if (y_true[i] != *label && y_pred[i] == *label) {
                    fp++;
                } else if (y_true[i] == *label && y_pred[i] != *label) {
                    fn++;
                } else {
                    tn++;
                }
            }
            float precision = (tp + fp) == 0 ? 0 : tp / static_cast<float>(tp + fp);
            float recall = (tp + fn) == 0 ? 0 : tp / static_cast<float>(tp + fn);
            float f1 = (precision + recall) == 0 ? 0 : 2 * (precision * recall) / (precision + recall);
            float accuracy = (tp + tn) / static_cast<float>(y_true.size());
            map<string, float> metrics;
            metrics["Precision"] = precision;
            metrics["Recall"] = recall;
            metrics["F1-score"] = f1;
            metrics["Accuracy"] = accuracy;
            report[*label] = metrics;
        }
        return report;
    }
};

// KNNClassifier class to implement K-Nearest Neighbors algorithm from scratch
class KNNClassifier {
private:
    int k;  // Number of nearest neighbors to consider
    vector<vector<float> > X_train;  // Training features
    vector<string> y_train;  // Training labels

public:
    // Constructor to initialize the number of neighbors
    explicit KNNClassifier(int k_val = 3) : k(k_val) {}

    // Store the training data
    void fit(const vector<vector<float> >& X, const vector<string>& y) {
        X_train = X;
        y_train = y;
    }

    // Predict the class of a single feature set
    string predict_single(const vector<float>& input_features) {
        vector<pair<float, string> > distances; // Store distances and corresponding labels

        // Calculate Euclidean distance from input features to all training samples
        for (size_t i = 0; i < X_train.size(); ++i) {
            float distance = 0.0f;
            for (size_t j = 0; j < X_train[i].size(); ++j) {
                distance += pow(X_train[i][j] - input_features[j], 2);
            }
            distance = sqrt(distance); // Square root of sum of squares
            distances.push_back(make_pair(distance, y_train[i])); // Store distance and label
        }

        // Sort distances
        sort(distances.begin(), distances.end());

        // Count the labels among the k nearest neighbors
        map<string, int> label_count;
        for (int i = 0; i < k && i < static_cast<int>(distances.size()); ++i) {
            label_count[distances[i].second]++;
        }

        // Determine the most frequent label
        string prediction;
        int max_count = -1; // Initialize to -1 to ensure any count will be higher initially
        for (map<string, int>::const_iterator label = label_count.begin(); label != label_count.end(); ++label) {
            if (label->second > max_count) {
                max_count = label->second;
                prediction = label->first;
            }
        }

        return prediction; // Return the most common label
    }

    // Predict a list of feature sets
    vector<string> predict(const vector<vector<float> >& X) {
        vector<string> predictions;
        for (vector<vector<float> >::const_iterator features = X.begin(); features != X.end(); ++features) {
            predictions.push_back(predict_single(*features));
        }
        return predictions;
    }

    // Method to calculate the accuracy of the predictions
    float accuracy(const vector<string>& y_test, const vector<string>& y_pred) {
        int correct = 0;
        for (size_t i = 0; i < y_test.size(); ++i) {
            if (y_test[i] == y_pred[i]) {
                correct++;
            }
        }
        return static_cast<float>(correct) / y_test.size();
    }

    // Classification report generation
    map<string, map<string, float> > classification_report(const vector<string>& y_true, const vector<string>& y_pred) {
        map<string, map<string, float> > report;
        map<string, int> true_counts, pred_counts, true_positive;

        for (size_t i = 0; i < y_true.size(); ++i) {
            true_counts[y_true[i]]++;
            pred_counts[y_pred[i]]++;
            if (y_true[i] == y_pred[i]) {
                true_positive[y_true[i]]++;
            }
        }

        for (map<string, int>::const_iterator label = true_counts.begin(); label != true_counts.end(); ++label) {
            float precision = (pred_counts[label->first] == 0) ? 0 : true_positive[label->first] / static_cast<float>(pred_counts[label->first]);
            float recall = true_positive[label->first] / static_cast<float>(label->second);
            float f1 = (precision + recall == 0) ? 0 : 2 * (precision * recall) / (precision + recall);
            float accuracy = static_cast<float>(true_positive[label->first]) / y_true.size();

            map<string, float> metrics;
            metrics["Precision"] = precision;
            metrics["Recall"] = recall;
            metrics["F1-score"] = f1;
            metrics["Accuracy"] = accuracy;

            report[label->first] = metrics;
        }

        return report;
    }
};

// Support Vector Machine Classifier
class SVMClassifier {
private:
    vector<float> weights;
    float bias;
    float learning_rate;
    float lambda_param;
    int n_iters;

public:
    // Constructor to initialize SVM parameters
    SVMClassifier(float lr = 0.001, float lambda = 0.01, int n = 1000) : learning_rate(lr), lambda_param(lambda), n_iters(n), bias(0.0) {}

    // Method to train the SVM classifier
    void fit(const vector<vector<float> >& X, const vector<string>& y) {
        int n_samples = X.size();
        int n_features = X[0].size();
        weights.resize(n_features, 0.0);

        vector<int> y_binary(n_samples);
        for (int i = 0; i < n_samples; ++i) {
            y_binary[i] = (y[i] == "Good") ? 1 : -1;
        }

        for (int _ = 0; _ < n_iters; ++_) {
            for (int i = 0; i < n_samples; ++i) {
                float condition = y_binary[i] * (dot_product(X[i], weights) + bias);
                if (condition >= 1) {
                    for (int j = 0; j < n_features; ++j) {
                        weights[j] -= learning_rate * (2 * lambda_param * weights[j]);
                    }
                } else {
                    for (int j = 0; j < n_features; ++j) {
                        weights[j] -= learning_rate * (2 * lambda_param * weights[j] - X[i][j] * y_binary[i]);
                    }
                    bias -= learning_rate * y_binary[i];
                }
            }
        }
    }

    // Method to predict the class of a single feature set
    string predict_single(const vector<float>& input_features) {
        float linear_output = dot_product(input_features, weights) + bias;
        return (linear_output >= 0) ? "Good" : "Bad";
    }

    // Method to predict a list of feature sets
    vector<string> predict(const vector<vector<float> >& X) {
        vector<string> predictions;
        for (vector<vector<float> >::const_iterator features = X.begin(); features != X.end(); ++features) {
            predictions.push_back(predict_single(*features));
        }
        return predictions;
    }

    // Method to calculate the accuracy of the predictions
    float accuracy(const vector<string>& y_test, const vector<string>& y_pred) {
        int correct = 0;
        for (size_t i = 0; i < y_test.size(); ++i) {
            if (y_test[i] == y_pred[i]) {
                correct++;
            }
        }
        return static_cast<float>(correct) / y_test.size();
    }

    // Classification report generation
    map<string, map<string, float> > classification_report(const vector<string>& y_true, const vector<string>& y_pred) {
        set<string> unique_labels(y_true.begin(), y_true.end());
        map<string, map<string, float> > report;
        for (set<string>::const_iterator label = unique_labels.begin(); label != unique_labels.end(); ++label) {
            int tp = 0, fp = 0, fn = 0, tn = 0;
            for (size_t i = 0; i < y_true.size(); ++i) {
                if (y_true[i] == *label && y_pred[i] == *label) {
                    tp++;
                } else if (y_true[i] != *label && y_pred[i] == *label) {
                    fp++;
                } else if (y_true[i] == *label && y_pred[i] != *label) {
                    fn++;
                } else {
                    tn++;
                }
            }
            float precision = (tp + fp) == 0 ? 0 : tp / static_cast<float>(tp + fp);
            float recall = (tp + fn) == 0 ? 0 : tp / static_cast<float>(tp + fn);
            float f1 = (precision + recall) == 0 ? 0 : 2 * (precision * recall) / (precision + recall);
            float accuracy = (tp + tn) / static_cast<float>(y_true.size());
            map<string, float> metrics;
            metrics["Precision"] = precision;
            metrics["Recall"] = recall;
            metrics["F1-score"] = f1;
            metrics["Accuracy"] = accuracy;
            report[*label] = metrics;
        }
        return report;
    }

private:
    // Helper function to calculate dot product
    float dot_product(const vector<float>& a, const vector<float>& b) {
        float result = 0.0;
        for (size_t i = 0; i < a.size(); ++i) {
            result += a[i] * b[i];
        }
        return result;
    }
};

// Function to print the classification report in a readable format
void print_classification_report(const map<string, map<string, float> >& report) {
    cout << "Classification Report:\n";
    for (map<string, map<string, float> >::const_iterator label_metrics = report.begin(); label_metrics != report.end(); ++label_metrics) {
        cout << "\nClass " << label_metrics->first << ":\n";
        for (map<string, float>::const_iterator metric = label_metrics->second.begin(); metric != label_metrics->second.end(); ++metric) {
            cout << "  " << metric->first << ": " << metric->second << "\n";
        }
    }
}

// Main function to handle user interaction, data processing, and model training/prediction
int main() {
    
    // Initialize DataHandler with a specific CSV file path
    DataHandler data("banana_quality.csv");

    // Classifier instances
    NaiveBayesClassifier nb;
    KNNClassifier knn(3);
    SVMClassifier svm;

    // Fit the classifiers with the training data
    nb.fit(data.features, data.labels);
    knn.fit(data.features, data.labels);
    svm.fit(data.features, data.labels);

    // Predict using the test data
    vector<string> nb_predictions = nb.predict(data.features);
    vector<string> knn_predictions = knn.predict(data.features);
    vector<string> svm_predictions = svm.predict(data.features);

    // Generate the classification reports
    map<string, map<string, float> > nb_report = nb.classification_report(data.labels, nb_predictions);
    map<string, map<string, float> > knn_report = knn.classification_report(data.labels, knn_predictions);
    map<string, map<string, float> > svm_report = svm.classification_report(data.labels, svm_predictions);

    // Print the classification reports
    cout << "Naive Bayes Classifier Report:\n";
    print_classification_report(nb_report);

    cout << "\nK-Nearest Neighbors Classifier Report:\n";
    print_classification_report(knn_report);

    cout << "\nSupport Vector Machine Classifier Report:\n";
    print_classification_report(svm_report);

    return 0;
}

File: Code.py
"""
Rosolino Mangano
CIS 3100
Professor Jairam
05/22/2024
"""

import csv
import random
import math
from random import shuffle

# DataHandler class to handle reading CSV file, splitting data into train/test sets, and separating features and labels
class DataHandler:
    def __init__(self, filepath):
        # Constructor to initialize the filepath
        self.filepath = filepath
        self.data = []

    def read_csv(self):
        # Read data from a CSV file and store it in a list
        with open(self.filepath, 'r') as file:
            csv_reader = csv.reader(file)
            self.header = next(csv_reader)  # Store the header row
            self.data = [row for row in csv_reader]
        print("Header Row:", self.header)  # Print the header row to verify the column names

    def train_test_split(self, test_size=0.2):
        # Shuffle the dataset to ensure randomness
        shuffle(self.data)
        # Determine the split index based on the test size
        split_index = int(len(self.data) * (1 - test_size))
        # Split the dataset into training and testing sets
        self.train_data = self.data[:split_index]
        self.test_data = self.data[split_index:]

    def separate_features_labels(self, target_column):
        # Find the index of the target column
        target_index = self.header.index(target_column)
        # Separate the features and labels for training and testing sets
        self.X_train = [list(map(float, row[:target_index] + row[target_index+1:])) for row in self.train_data]
        self.y_train = [row[target_index] for row in self.train_data]
        self.X_test = [list(map(float, row[:target_index] + row[target_index+1:])) for row in self.test_data]
        self.y_test = [row[target_index] for row in self.test_data]

# NaiveBayesClassifier class to implement Naive Bayes algorithm from scratch
class NaiveBayesClassifier:
    def __init__(self):
        # Initialize dictionaries to store the means, standard deviations,
        # and class probabilities for each class
        self.means = {}
        self.stds = {}
        self.class_probabilities = {}

    def fit(self, X, y):
        # Train the classifier by calculating the class probabilities
        # and the means and standard deviations for each feature
        self._calculate_class_probabilities(y)
        self._calculate_means_stds(X, y)

    def _calculate_class_probabilities(self, y):
        # Calculate the probability of each class based on label frequency
        class_counts = {label: y.count(label) for label in set(y)}
        total_count = len(y)
        self.class_probabilities = {label: count / total_count for label, count in class_counts.items()}

    def _calculate_means_stds(self, X, y):
        # Calculate the mean and standard deviation for each class and each feature
        for label in self.class_probabilities:
            # Extract features for instances of the current class
            label_features = [X[i] for i in range(len(X)) if y[i] == label]
            # Calculate mean and standard deviation for each feature
            self.means[label] = [sum(f) / len(f) for f in zip(*label_features)]
            self.stds[label] = [math.sqrt(sum([(x - mean)**2 for x in f]) / len(f)) for mean, f in zip(self.means[label], zip(*label_features))]

    def predict_single(self, input_features):
        # Predict the class of a single feature set
        probabilities = {}
        for label, _ in self.means.items():
            # Start with the prior probability of the class
            probabilities[label] = self.class_probabilities[label]
            # Multiply by the probability of each feature
            for i, feature in enumerate(input_features):
                probabilities[label] *= self._calculate_probability(feature, self.means[label][i], self.stds[label][i])
        # Return the class with the highest probability
        return max(probabilities, key=probabilities.get)

    def _calculate_probability(self, x, mean, std):
        # Calculate the probability of a feature value with a Gaussian distribution
        exponent = math.exp(-(math.pow(x-mean,2)/(2*math.pow(std,2))))
        return (1 / (math.sqrt(2*math.pi) * std)) * exponent

    def predict(self, X):
        # Predict a list of feature sets
        return [self.predict_single(features) for features in X]

    def accuracy(self, y_test, y_pred):
        # Calculate the accuracy of the predictions
        correct = sum(1 for true, pred in zip(y_test, y_pred) if true == pred)
        return correct / len(y_test)

    def classification_report(self, y_true, y_pred):
        # Generate a classification report for the predictions
        unique_labels = set(y_true)
        report = {}
        for label in unique_labels:
            tp = sum(1 for i in range(len(y_true)) if y_true[i] == label and y_pred[i] == label)
            fp = sum(1 for i in range(len(y_true)) if y_true[i] != label and y_pred[i] == label)
            fn = sum(1 for i in range(len(y_true)) if y_true[i] == label and y_pred[i] != label)
            tn = sum(1 for i in range(len(y_true)) if y_true[i] != label and y_pred[i] != label)

            # Calculate precision, recall, and F1-score for each class
            precision = tp / (tp + fp) if tp + fp > 0 else 0
            recall = tp / (tp + fn) if tp + fn > 0 else 0
            f1 = 2 * (precision * recall) / (precision + recall) if precision + recall > 0 else 0
            accuracy = (tp + tn) / len(y_true)

            report[label] = {
                'Precision': precision,
                'Recall': recall,
                'F1-score': f1,
                'Accuracy': accuracy
            }

        return report

# KNNClassifier class to implement K-Nearest Neighbors algorithm from scratch
class KNNClassifier:
    def __init__(self, k=3):
        # Initialize KNN with a specified number of neighbors (k)
        self.k = k

    def fit(self, X_train, y_train):
        # Store the training data
        self.X_train = X_train
        self.y_train = y_train

    def predict(self, X_test):
        # Predict the class labels for the test data
        return [self.predict_single(x) for x in X_test]

    def predict_single(self, input_features):
        # Predict the class label for a single feature set
        distances = []
        for i in range(len(self.X_train)):
            distance = self.euclidean_distance(input_features, self.X_train[i])
            distances.append((distance, self.y_train[i]))
        distances.sort(key=lambda x: x[0])  # Sort by distance
        k_nearest_neighbors = distances[:self.k]  # Get the k nearest neighbors
        labels = [label for _, label in k_nearest_neighbors]
        prediction = max(set(labels), key=labels.count)  # Return the most common label
        return prediction

    def euclidean_distance(self, point1, point2):
        # Calculate the Euclidean distance between two points
        return math.sqrt(sum((x - y) ** 2 for x, y in zip(point1, point2)))

    def accuracy(self, y_test, y_pred):
        # Calculate the accuracy of the predictions
        correct = sum(1 for true, pred in zip(y_test, y_pred) if true == pred)
        return correct / len(y_test)

    def classification_report(self, y_true, y_pred):
        # Generate a classification report for the predictions
        unique_labels = set(y_true)
        report = {}
        for label in unique_labels:
            tp = sum(1 for i in range(len(y_true)) if y_true[i] == label and y_pred[i] == label)
            fp = sum(1 for i in range(len(y_true)) if y_true[i] != label and y_pred[i] == label)
            fn = sum(1 for i in range(len(y_true)) if y_true[i] == label and y_pred[i] != label)
            tn = sum(1 for i in range(len(y_true)) if y_true[i] != label and y_pred[i] != label)

            # Calculate precision, recall, and F1-score for each class
            precision = tp / (tp + fp) if tp + fp > 0 else 0
            recall = tp / (tp + fn) if tp + fn > 0 else 0
            f1 = 2 * (precision * recall) / (precision + recall) if precision + recall > 0 else 0
            accuracy = (tp + tn) / len(y_true)

            report[label] = {
                'Precision': precision,
                'Recall': recall,
                'F1-score': f1,
                'Accuracy': accuracy
            }

        return report

# SVMClassifier class to implement a simple linear Support Vector Machine from scratch
class SVMClassifier:
    def __init__(self, learning_rate=0.001, lambda_param=0.01, n_iters=1000):
        # Initialize SVM with specified parameters
        self.lr = learning_rate
        self.lambda_param = lambda_param
        self.n_iters = n_iters
        self.w = None
        self.b = None

    def fit(self, X_train, y_train):
        # Convert labels to -1 or 1
        y_train = [1 if label == 'Good' else -1 for label in y_train]  # Assuming 'Good' and 'Bad' labels
        n_samples, n_features = len(X_train), len(X_train[0])

        # Initialize weights and bias
        self.w = [0.0] * n_features
        self.b = 0.0

        # Perform gradient descent
        for _ in range(self.n_iters):
            for idx, x_i in enumerate(X_train):
                condition = y_train[idx] * (sum(self.w[i] * x_i[i] for i in range(n_features)) + self.b) >= 1
                if condition:
                    self.w = [self.w[i] - self.lr * (2 * self.lambda_param * self.w[i]) for i in range(n_features)]
                else:
                    self.w = [self.w[i] - self.lr * (2 * self.lambda_param * self.w[i] - x_i[i] * y_train[idx]) for i in range(n_features)]
                    self.b -= self.lr * y_train[idx]

    def predict(self, X_test):
        # Predict the class labels for the test data
        return [self.predict_single(x) for x in X_test]

    def predict_single(self, input_features):
        # Predict the class label for a single feature set
        linear_output = sum(self.w[i] * input_features[i] for i in range(len(input_features))) + self.b
        return 'Good' if linear_output >= 0 else 'Bad'

    def accuracy(self, y_test, y_pred):
        # Calculate the accuracy of the predictions
        correct = sum(1 for true, pred in zip(y_test, y_pred) if true == pred)
        return correct / len(y_test)

    def classification_report(self, y_true, y_pred):
        # Generate a classification report for the predictions
        unique_labels = set(y_true)
        report = {}
        for label in unique_labels:
            tp = sum(1 for i in range(len(y_true)) if y_true[i] == label and y_pred[i] == label)
            fp = sum(1 for i in range(len(y_true)) if y_true[i] != label and y_pred[i] == label)
            fn = sum(1 for i in range(len(y_true)) if y_true[i] == label and y_pred[i] != label)
            tn = sum(1 for i in range(len(y_true)) if y_true[i] != label and y_pred[i] != label)

            # Calculate precision, recall, and F1-score for each class
            precision = tp / (tp + fp) if tp + fp > 0 else 0
            recall = tp / (tp + fn) if tp + fn > 0 else 0
            f1 = 2 * (precision * recall) / (precision + recall) if precision + recall > 0 else 0
            accuracy = (tp + tn) / len(y_true)

            report[label] = {
                'Precision': precision,
                'Recall': recall,
                'F1-score': f1,
                'Accuracy': accuracy
            }

        return report

# Function to print the classification report in a readable format
def print_classification_report(report):
    print("Classification Report:")
    for label, metrics in report.items():
        print(f"\nClass {label}:")
        for metric, value in metrics.items():
            print(f"  {metric}: {value:.2f}")

# Function to select and return the classifier based on user input
def select_classifier():
    while True:
        print("\nSelect the classifier:")
        print("1: Naive Bayes")
        print("2: K-Nearest Neighbors")
        print("3: Support Vector Machine")
        choice = input("Enter the number of your choice: ")

        if choice == '1':
            return NaiveBayesClassifier()
        elif choice == '2':
            return KNNClassifier(k=3)  # You can adjust the value of k
        elif choice == '3':
            return SVMClassifier()
        else:
            print("Invalid choice. Please select again.")

# Main function to handle user interaction, data processing, and model training/prediction
def main():

    # Defines the path to the CSV file containing the dataset
    filepath = 'banana_quality.csv'
    
    # Initialize DataHandler
    data_handler = DataHandler(filepath)
    data_handler.read_csv()
    data_handler.train_test_split(test_size=0.2)
    data_handler.separate_features_labels(target_column='Quality')

    while True:
        # User selects which classifier to use
        classifier = select_classifier()

        # Fit the classifier
        classifier.fit(data_handler.X_train, data_handler.y_train)

        # Predict and evaluate
        y_pred = classifier.predict(data_handler.X_test)
        accuracy = classifier.accuracy(data_handler.y_test, y_pred)
        report = classifier.classification_report(data_handler.y_test, y_pred)

        # Print accuracy and classification report
        print(f"\nAccuracy: {accuracy:.4f}")
        print_classification_report(report)

        # Simple User Interface for predictions
        while True:
            user_input = input("\nEnter features separated by commas or 'exit' to quit or 'change' to select a new classifier: ")
            if user_input.lower() == 'exit':
                return
            elif user_input.lower() == 'change':
                break
            try:
                input_features = [float(x) for x in user_input.split(',')]
                prediction = classifier.predict_single(input_features)
                print(f"Prediction: {prediction}")
            except ValueError:
                print("Invalid input. Please enter numerical feature values separated by commas.")

if __name__ == "__main__":
    main()
